_posts/2018-10-22-release-1-5.md

```diff
@@ -0,0 +1,199 @@
+release, and upcoming [1.6](https://github.com/pmem/issues/issues/932), are
+already using this new open process. When you enter one of the release, you will
+see that most of the planned features have their own pages where they can be 
+discussed.
+
+The goal of this change is to keep our users informed about future plans, but
+also to encourage discussion with the community about the direction we are taking.
+We'd like invite everyone interested in PMDK to start directly influencing
+and contributing to our future work.
+
+Since PMDK project is still relatively small, this process isn't formalized.
+Think we are missing a feature in a library? Write a comment on the release
+planning page or create a new issue on our [tracker](https://github.com/pmem/issues/issues?q=is%3Aopen+is%3Aissue+label%3A%22Type%3A+Feature%22),
+that's it.
+
+I can promise that no feature request will get unanswered, at least for now ;)
+
+# PMDK 1.5
+
+This is a *big* release, with many quality of life enhancements, new features
+and performance improvements. We've also introduced an entirely new consistency
+checking tool to compliment pmemcheck.
+
+To keep this relatively short, here I'll just briefly describe each item, and
+the bigger ones will get their own posts describing the details of the change.
+
+## Reliability, Availability and Serviceability (RAS)
+
+NVDIMMs are storage and as such have to deal with on-media errors and hardware
+failures. But they are also memory, which means that an application can
+encounter poisoned memory pages at runtime, and what's more, those poisoned
+pages persist across restarts and need to be handled manually.
+
+To account for that, we've worked on two main RAS features:
+
+### Unsafe (dirty) shutdown count
+
+This is used to detect if an ADR, feature on which persistent memory programming
+model relies on to avoid having to flush the memory controller caches, failure
+has occurred while a pool has been opened. Based on this information, the
+library can determine whether the pool could have been corrupted or not.
+
+### Bad blocks support
+
+On Linux, if application maps a persistent memory resident file with
+a bad block, everything will work correctly up until the bad, poisoned, page
+is accessed for the first time in the current instance of the application. That
+first page fault will cause a SIGBUS and the process will most likely be
+terminated.
+
+To prevent this cycle from happening over and over again, PMDK will now detect
+if the files that the pool is composed of have any badblocks, and if so, it will
+refuse to open it. This allows the user to handle the failure gracefully.
+
+In addition to that, we implemented recovery of bad blocks from poolset
+replicas.
+
+## pmemreorder
+
+Almost since the beginning of this project, we planned on implementing a tool
+which could exhaustively check consistency for all possible combination of
+persistent memory stores based on runtime binary instrumentation.
```

Comment:
stores->flushes?

Reply:
Im my opinion  `persistent memory stores` is fine.

Reply:
Stores are made by applications in very specific order. But they don't control in which order dirty cache lines are evicted (and they can be evicted even before explicit flush), so what we are simulating with pmreorder is reordering of flushes. Of course from perspective of the application this will look like CPU reordered its stores, but saying that pmreorder "checks consistency for all possible combination of persistent memory stores" sounds like solution to a non-existent problem, because application doesn't make stores in random order.

Reply:
I disagree. Application might not store data in random order, but the CPU might actually move data to the NVDIMM in a fashion that is not-deterministic from the point of view the application.
While yes, you are technically speaking correct that we reorder "flushes from the CPU cache", but what I wrote is also correct if you don't assume that the word 'store' means a MOV instruction, but in addition to that, I think it's more understandable for regular people.

I changed it to "stores to the NVDIMM", to make it clear than I'm mentally bypassing the CPU cache.